drivers/hwmon_ina219.c

```diff
@@ -0,0 +1,339 @@
+/* hwmon.c Driver for INA219 hwmon-based power monitors.
+
+	Copyright (C) 2024 Jan Viktorin <jan.viktorin@gmail.com>
+
+	This program is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with this program; if not, write to the Free Software
+	Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+#include "main.h"
+
+#include <dirent.h>
+#include <errno.h>
+#include <string.h>
+#include <sys/types.h>
+
+#ifndef STRFY0
+	#define STRFY0(x) #x
+#endif
+#ifndef STRFY
+	#define STRFY(x) STRFY0(x)
+#endif
+
+#define SYSFS_HWMON_DIR                     "/sys/class/hwmon"
+#define LOW_BATTERY                         15
+#define HIGH_BATTERY                        75
+#define DRIVER_NAME                         "hwmon-INA219 UPS driver"
+#define DRIVER_VERSION                      "0.1"
+
+upsdrv_info_t upsdrv_info = {
+	DRIVER_NAME,
+	DRIVER_VERSION,
+	"Jan Viktorin <jan.viktorin@gmail.com>",
+	DRV_EXPERIMENTAL,
+	{ NULL },
+};
+
+/**
+ * @brief Path usually pointing to /sys/class/hwmon/hwmonX.
+ */
+static char ina219_base_path[PATH_MAX];
+
+/**
+ * @brief Threshold for detection of LB status.
+ */
+static unsigned int low_battery = LOW_BATTERY;
+
+/**
+ * @brief Threshold for detection of HB status.
+ */
+static unsigned int high_battery = HIGH_BATTERY;
+
+/**
+ * @brief Voltage value as recently read from in1_input.
+ */
```

Comment:
Makes sense to comment that it would (may?) be in milliVolts / milliAmps (below) so you would shift it by 3 orders of magnitude into standard units range.

In fact, is it "guaranteed" that the values from sysfs would be integers and in milliUnits? Would it make sense to parse them in `file_read_number()` as e.g. `double` and range-check (if over 1000 then shift else use as is)? Just something to think about, I have no answers here :)

Reply:
> Makes sense to comment that it would (may?) be in milliVolts / milliAmps (below) so you would shift it by 3 orders of magnitude into standard units range.

Mentioned in the doc comment.

> In fact, is it "guaranteed" that the values from sysfs would be integers and in milliUnits? Would it make sense to parse them in file_read_number() as e.g. double and range-check (if over 1000 then shift else use as is)? Just something to think about, I have no answers here :)

Yes. From kernel doc, it is clearly stated that the miliUnits are used [1]. In sysfs interface standard [2], it is stated that "All sysfs values are fixed point numbers." You can find the print format "%d" in the kernel source code [3].

[1] https://docs.kernel.org/hwmon/ina2xx.html#general-sysfs-entries
[2] https://docs.kernel.org/hwmon/sysfs-interface.html#naming-and-data-format-standards-for-sysfs-files
[3] https://github.com/torvalds/linux/blob/7367539ad4b0f8f9b396baf02110962333719a48/drivers/hwmon/ina2xx.c#L300C16-L300C33

Reply:
Thanks! Maybe it's worth pointing out in code comments, on one hand to avoid similar questions in the future, and on another to help eventual portability (whether to some newer Linux kernel 10.x convention or perhaps to extend this driver to more OSes with concepts similar to `hwmon`).

Reply:
I added few links there.